### PR TITLE
Molecular surfaces (marching cubes)

### PR DESCRIPTION
**Pull request by rlatawiec**
A module for generating, plotting, and computation on molecular surfaces (for now only using marching cubes).

### CODE REVIEW

File: oddt/surface.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    for ellip, coord_min, coord_max in zip(ellip_list, ellip_coord_min, ellip_coord_max):
+        grid[coord_min[0]:coord_max[0],
+             coord_min[1]:coord_max[1],
+             coord_min[2]:coord_max[2]] += ellip
+    spacing = (1 / scaling,) * 3
+    # Hole-filling with morphological closing
+    grid = binary_closing(grid, ball(probe_radius * 2 * scaling))
+    # Marching cubes
+    verts, faces, normals, values = measure.marching_cubes_lewiner(grid, level=0, step_size=1, spacing=spacing)
+    return verts - offset / scaling, faces
```
Comment (marta-sd) [inline oddt/surface.py]: I think this part requires a comment - at first I thought there are missing parentheses around `verts - offset`. Point out that `verts` are already in the orignal scale.

File: oddt/surface.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+def draw_surface_marching_cubes(molecule, scaling=1, plot_atoms=False, alpha=1, color=(0, 0, 1),
+                                figure_size=(10, 10), remove_hoh=False, probe_radius=1.4):
+    """Plots a surface of a molecule in 3D.
+
+    Parameters
```
Comment (marta-sd) [inline oddt/surface.py]: I personally prefer simple functions performing simple tasks. Consider using vertices and faces as arguments instead of parameters required to generate the surface. When we define plotting function like this, it will be more general and robust. Now we can't use it to plot surfaces generated with different methods (we'll need to write second, redundant function). Also, if we for example add a new parameter to `generate_surface_marching_cubes` it might break the drawing function.

File: oddt/surface.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    # Copy and pre-process the molecule
+    mol = molecule.clone
+    mol.removeh()
+    if remove_hoh:
+        mol.protein = True
+        no_hoh = mol.atom_dict['resname'] != 'HOH'
+        atom_dict = mol.atom_dict[no_hoh]
+    else:
+        atom_dict = mol.atom_dict
+    # Take a molecule's coordinates and atom radii and scale if necessary
```
Comment (marta-sd) [inline oddt/surface.py]: Separate blocks of code with empty lines, it will be easier to read.

File: oddt/surface.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        by a factor of scaling. Results in a more accurate representation of the surface,
+        and therefore more accurate computation of distances but increases computation time.
+
+    Returns
+    -------
+    atom_dict : numpy array
+        An atom_dict containing only the surface residues from the original molecule.
+    """
+    # Copy the atom_dict, remove waters
+    atom_dict = molecule.atom_dict.copy()
```
Comment (marta-sd) [inline oddt/surface.py]: Why do we need to copy `atom_dict`? If there is a reason - write it in the comment.

File: oddt/surface.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    no_hoh = atom_dict['resname'] != 'HOH'
+    atom_dict = atom_dict[no_hoh]
+    coords = atom_dict['coords']
+    if not max_dist:
```
Comment (marta-sd) [inline oddt/surface.py]: This part needs a comment, it's not clear that `max_dist` is a vector.

**Comment by mwojcikowski** [comment]
Please review Codacy checks.

File: oddt/surface.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+"""This module generates, plots, and does computation with molecular surfaces.
+"""
+
+import oddt
+import numpy as np
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d.art3d import Poly3DCollection
```
Comment (mwojcikowski) [inline oddt/surface.py]: Unused imports

File: oddt/surface.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+from skimage import measure
+from skimage.morphology import ball, binary_closing
+from scipy.spatial.distance import cdist
+
+
+def generate_surface_marching_cubes(molecule, remove_hoh=False, scaling=1, probe_radius=1.4):
+    """Generates a molecular surface mesh using the marching_cubes_lewiner method from scikit-image.
```
Comment (mwojcikowski) [inline oddt/surface.py]: I think those lines could be a bit shorter. I don't see why we shouldn't adhere to PEP8

File: oddt/surface.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    faces : numpy array
+        Faces are defined by referencing vertices from verts.
+    """
+    # Copy and pre-process the molecule
+    mol = molecule.clone
+    mol.removeh()
+    if remove_hoh:
+    # Residue names are needed for water removal, molecule.protein property must be set to True"
+        mol.protein = True
```
Comment (mwojcikowski) [inline oddt/surface.py]: Check if Molecule is protein, dont set it here.

File: oddt/surface.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    -------
+    verts : numpy array
+        Spatial coordinates for mesh vertices.
+
+    faces : numpy array
+        Faces are defined by referencing vertices from verts.
+    """
+    # Copy and pre-process the molecule
+    mol = molecule.clone
+    mol.removeh()
```
Comment (mwojcikowski) [inline oddt/surface.py]: Dont clone nor remove Hs here. Instead remove them from atom_dict

File: oddt/surface.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    if not max_dist:
+        max_dist = atom_dict['radius']
+
+    # Marching cubes
+    verts, _ = generate_surface_marching_cubes(molecule, remove_hoh=True,
+                                               scaling=scaling, probe_radius=1.4)
+
+    # Calculate distances between atoms and the surface
+    tree_verts = cKDTree(verts)
+    mask = [tree_verts.query_ball_point(point, radius)
```
Comment (mwojcikowski) [inline oddt/surface.py]: Instead of two loops over mask wrap the `query_ball_point` with `bool` 

`bool(tree_verts.query_ball_point(...))`

File: oddt/surface.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                                    probe_radius=1.4):
+    """Generates a molecular surface mesh using the marching_cubes_lewiner
+    method from scikit-image.
+
+    Parameters
+    ----------
+    molecule : oddt.toolkit.Molecule object
+        Molecule for which the surface will be generated
+
+    remove_hoh : bool (default = False)
```
Comment (marta-sd) [inline oddt/surface.py]: User needs to know that `molecule.protein` must be set to True in order to use this option. Add this requirement to docstring.

File: oddt/surface.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    Returns
+    -------
+    verts : numpy array
+        Spatial coordinates for mesh vertices.
+
+    faces : numpy array
+        Faces are defined by referencing vertices from verts.
+    """
+    atom_dict = molecule.atom_dict
```
Comment (marta-sd) [inline oddt/surface.py]: Add input validation please - we must be sure that each argument has a correct type and value (raise `TypeError` or `ValueError` with a meaningful message if input is incorrect).

File: oddt/surface.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+             coord_min[1]:coord_max[1],
+             coord_min[2]:coord_max[2]] += ball_dict[radius]
+    spacing = (1 / scaling,) * 3
+
+    # Hole-filling with morphological closing
+    grid = binary_closing(grid, ball(probe_radius * 2 * scaling))
+
+    # Marching cubes
+    verts, faces, _, _ = measure.marching_cubes_lewiner(grid, level=0, step_size=1, spacing=spacing)
+    return verts - offset / scaling, faces
```
Comment (marta-sd) [inline oddt/surface.py]: I asked you to add a comment here in the previous review: [discussion_r133442155](http://github.com/oddt/oddt/pull/37#discussion_r133442155)

File: oddt/surface.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+
+def find_surface_residues(molecule, max_dist=None, scaling=1):
+    """Finds residues close to the molecular surface using
+    generate_surface_marching_cubes.
+
+    Parameters
+    ----------
+    molecule : oddt.toolkit.Molecule
```
Comment (marta-sd) [inline oddt/surface.py]: Add information that `molecule.protein` must be set to True

File: oddt/surface.py [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        max_dist = atom_dict['radius']
+
+    # Marching cubes
+    verts, _ = generate_surface_marching_cubes(molecule, remove_hoh=True,
+                                               scaling=scaling, probe_radius=1.4)
+
+    # Calculate distances between atoms and the surface
+    tree_verts = cKDTree(verts)
+    mask = [bool(tree_verts.query_ball_point(point, radius))
+            for point, radius in zip(coords, max_dist)]
```
Comment (marta-sd) [inline oddt/surface.py]: I'm afraid this zip doesn't work when `max_dist` is a single number.

File: oddt/surface.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    atom_dict : numpy array
+        An atom_dict containing only the surface residues
+        from the original molecule.
+    """
+    # Copy the atom_dict, remove waters
+    atom_dict = molecule.atom_dict
+    mask = (atom_dict['resname'] != 'HOH') & (atom_dict['atomicnum'] != 1)
+    atom_dict = atom_dict[mask]
+    coords = atom_dict['coords']
+    if not max_dist:
```
Comment (marta-sd) [inline oddt/surface.py]: I asked for a comment: [discussion_r133447051](http://github.com/oddt/oddt/pull/37#discussion_r133447051)

File: oddt/surface.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    ----------
+    molecule : oddt.toolkit.Molecule
+        Molecule to find surface residues in.
+
+    max_dist : array_like, numeric or None (default = None)
+        Maximum distance from the surface where residues would
+        still be considered close. If None, compares distances
+        to radii of respective atoms.
+
+    scaling : float (default = 1)
```
Comment (mwojcikowski) [inline oddt/surface.py]: `scaling` is int by default

File: oddt/surface.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+    Returns
+    -------
+    atom_dict : numpy array
+        An atom_dict containing only the surface residues
+        from the original molecule.
+    """
+    # Copy the atom_dict, remove waters
+    atom_dict = molecule.atom_dict
+    mask = (atom_dict['resname'] != 'HOH') & (atom_dict['atomicnum'] != 1)
```
Comment (mwojcikowski) [inline oddt/surface.py]: There is no hint in docstrings this (HOH removal) happen. We should at least mention it.

File: tests/test_surface.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+import oddt
+from oddt.surface import (generate_surface_marching_cubes,
+                          find_surface_residues)
+from sklearn.utils.testing import (assert_array_equal,
+                                   assert_array_almost_equal,
+                                   assert_equal,
+                                   assert_almost_equal,
+                                   assert_in)
+import pandas as pd
```
Comment (marta-sd) [inline tests/test_surface.py]: Unused import

File: tests/test_surface.py [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+test_data_dir = os.path.dirname(os.path.abspath(__file__))
+protein = next(oddt.toolkit.readfile('pdb', os.path.join(
+    test_data_dir, 'data/pdbbind/10gs_protein.pdb')))
+protein.protein = True
+protein.addh(only_polar=True)
+
+def test_generate_surface_marching_cubes():
+    """Tests generating surfaces"""
+    verts1, faces1 = generate_surface_marching_cubes(protein, scaling=1., probe_radius=1.4)
```
Comment (marta-sd) [inline tests/test_surface.py]: It's a good practice to specify all arguments (the ones with default values as well). This way you have bigger control over your tests. If we decide to change default value, the function still works but your test will fail.

Also you should provide some tests for all arguments:
- [ ] `remove_hoh` - add at least one test case with water removal. You can e.g. check verts and faces shape, like you do for `scaling`. I think it is also a good idea to check what happens if we use molecule without water residues and `remove_hoh=True` (expected behavior: water removal shouldn't change the surface)
- [ ] `probe_radius` - test different radii, especially 0.

File: tests/test_surface.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+    assert_equal(ref_vert_shape_1, verts1.shape)
+    assert_equal(ref_face_shape_1, faces1.shape)
+
+    assert_equal(ref_vert_shape_2, verts2.shape)
+    assert_equal(ref_face_shape_2, faces2.shape)
+
+    assert_equal(ref_vert_shape_3, verts3.shape)
+    assert_equal(ref_face_shape_3, faces3.shape)
+
```
Comment (marta-sd) [inline tests/test_surface.py]: I think you should also check if input validation works fine, you can use `assert_raises` from `nose.tools`.

File: tests/test_surface.py [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+def test_find_surface_residues():
+    """Tests finding residues on the surface"""
+    atom_dict_0 = find_surface_residues(protein, max_dist=0)
```
Comment (marta-sd) [inline tests/test_surface.py]: You should explicitly specify value for `scaling`. Or, even better, test different values - we can't be sure if it's actually passed to `generate_surface_marching_cubes`.

File: tests/test_surface.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    atom_dict_1 = find_surface_residues(protein, max_dist=2)
+    atom_dict_2 = find_surface_residues(protein, max_dist=3)
```
Comment (marta-sd) [inline tests/test_surface.py]: You should also test `max_dist=None`

File: tests/test_surface.py [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+    assert_equal(atom_dict_0.size, 0)
+    assert len(atom_dict_2) >= len(atom_dict_1) > len(atom_dict_0), ('Increasing max_dist should '
```
Comment (marta-sd) [inline tests/test_surface.py]: This should be written as two separate assertions - it's easier to debug your code when you precisely know which test failed.

File: oddt/surface.py [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+
+    # Create the grid
+    grid = np.zeros(shape=ball_coord_max.max(axis=0) + int(8 * scaling), dtype=bool)
+
+    # Place balls in grid
+    for radius, coord_min, coord_max in zip(radii, ball_coord_min, ball_coord_max):
+        grid[coord_min[0]:coord_max[0],
+             coord_min[1]:coord_max[1],
+             coord_min[2]:coord_max[2]] += ball_dict[radius]
+    spacing = (1 / scaling,) * 3
```
Comment (mwojcikowski) [inline oddt/surface.py]: Is that Py2 safe?

File: oddt/surface.py [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+        Expands the grid in which computation is done by
+        generate_surface_marching_cubes by a factor of scaling.
+        Results in a more accurate representation of the surface,
+        and therefore more accurate computation of distances
+        but increases computation time.
+
+    Returns
+    -------
+    atom_dict : numpy array
+        An atom_dict containing only the surface residues
```
Comment (mwojcikowski) [inline oddt/surface.py]: `atom_dict` contains atoms

File: tests/test_surface.py [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+protein = next(oddt.toolkit.readfile('pdb', os.path.join(
+    test_data_dir, 'data/dude/xiap/receptor_rdkit.pdb')))
+protein.protein = True
+protein.addh(only_polar=True)
+
+def test_generate_surface_marching_cubes():
+    """Tests generating surfaces"""
+    verts1, faces1 = generate_surface_marching_cubes(protein, scaling=1., probe_radius=1.4, remove_hoh=False)
+    verts2, faces2 = generate_surface_marching_cubes(protein, scaling=2., probe_radius=1.4, remove_hoh=False)
+    verts3, faces3 = generate_surface_marching_cubes(protein, scaling=1., probe_radius=1.4, remove_hoh=True)
```
Comment (mwojcikowski) [inline tests/test_surface.py]: I think the protein has no HOHs

File: oddt/surface.py [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+
+    # Marching cubes
+    verts, faces = measure.marching_cubes(grid, level=0, spacing=spacing)[:2]
+
+    # Verts already scaled by the marching cubes function (spacing parameter)
+    # Only need to scale the offset
+    return verts - offset / scaling, faces
+
+
+def find_surface_residues(molecule, max_dist=None, scaling=1.):
```
Comment (mwojcikowski) [inline oddt/surface.py]: The name of this function is misleading in my opinion. It finds atoms, not residues. Residues are indirectly found this way, as there are `res*` properties for each atom in `atom_dict`.

File: oddt/surface.py (hunk 25 above)
Comment (marta-sd) [reply oddt/surface.py]: I think it is - Rufin imports `division` from `__future__`

File: tests/test_surface.py (hunk 27 above)
Comment (marta-sd) [reply tests/test_surface.py]: I think we should:
* check whether `remove_hoh=True` changes anything (it shouldn't)
* test this function on a structure with some water residues

File: oddt/surface.py (hunk 25 above)
Comment (mwojcikowski) [reply oddt/surface.py]: Sure, I didn't notice that.
